pykafka/producer.py

```diff
@@ -317,30 +344,12 @@ def _get_partition_msgs(partition_id, req):
                 time.sleep(self._retry_backoff_ms / 1000)
                 self._prepare_request(to_retry, owned_broker, attempt)
             else:
-                raise ProduceFailureError('Unable to produce messages. See log for details.')
-
-    def _update_leaders(self):
-        """Ensure each message in each queue is in the queue owned by its
-            partition's leader
-
-        This function empties all broker queues, maps their messages to the
-        current leaders for their partitions, and enqueues the messages in
-        the appropriate queues.
-        """
-        # empty queues and figure out updated partition leaders
-        new_queue_contents = defaultdict(list)
-        for owned_broker in itervalues(self._owned_brokers):
-            owned_broker.lock.acquire()
-            current_queue_contents = owned_broker.flush(0, release_pending=True)
-            for kv, partition_id in current_queue_contents:
-                partition_leader = self._topic.partitions[partition_id].leader
-                new_queue_contents[partition_leader.id].append((kv, partition_id))
-        # retain locks for all brokers between these two steps
-        for owned_broker in itervalues(self._owned_brokers):
-            owned_broker.enqueue(new_queue_contents[owned_broker.broker.id],
-                                 self._block_on_queue_full)
-            owned_broker.resolve_event_state()
-            owned_broker.lock.release()
+                log.error("Failed to produce messages to broker %s:%s after %s attempts. "
+                          "Re-enqueuing %s messages.", owned_broker.broker.host,
+                          owned_broker.broker.port, attempt, len(to_retry))
+                owned_broker.increment_messages_pending(-1 * len(to_retry))
+                for tup in to_retry:
+                    self._produce(tup)
```

Comment:
I agree. The problem is that this branch is meant to make the producer not crash when a broker fails. If we raise an exception here, the client will need to handle that exception to avoid a crash. Maybe that's ok, but it at least complicates the code a client needs to write:

```python
with Producer() as p:
    try:
        p.produce(message)
    except ThisException:
        # what goes here? p.produce(message)?
```